src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -5287,12 +5287,16 @@ protected synchronized void drain(boolean isFinalShutdown) throws IOException, I
             CommitLog.instance.shutdownBlocking();
 
             // wait for miscellaneous tasks like sstable and commitlog segment deletion
-            ScheduledExecutors.nonPeriodicTasks.shutdown();
-            if (!ScheduledExecutors.nonPeriodicTasks.awaitTermination(1, MINUTES))
-                logger.warn("Unable to terminate non-periodic tasks within 1 minute.");
```

Comment:
So this log message would now appear as a TimeoutException in the main catch block on L5301?

Reply:
Yes. With one caveat and that is that instead of waiting 1 minute _for each executor_ (here we had just one), we are waiting 1 minute _for all of them_ to be shutdown / terminated. I consider this to be the flaw of the previous solution. They were not covering all other executors and I think that if whoever did that was taking other executors into account, he would end up with same solution.